velox/functions/prestosql/aggregates/MergeQDigestAggregate.h

```diff
@@ -0,0 +1,223 @@
+struct QDigestAccumulator {
+  explicit QDigestAccumulator(HashStringAllocator* allocator)
+      : digest_(
+            StlAllocator<T>(allocator),
+            facebook::velox::functions::qdigest::kUninitializedMaxError) {}
+
+  void mergeWith(StringView serialized) {
+    if (serialized.empty()) {
+      return;
+    }
+    digest_.mergeSerialized(serialized.data());
+  }
+
+  int64_t serializedSize() {
+    return digest_.serializedByteSize();
+  }
+
+  void serialize(char* outputBuffer) {
+    // QDigest from qdigest_agg() should always have maxError set
+    VELOX_USER_CHECK_NE(
+        digest_.getMaxError(),
+        facebook::velox::functions::qdigest::kUninitializedMaxError,
+        "QDigest maxError is uninitialized");
+    digest_.serialize(outputBuffer);
+  }
+
+ private:
+  facebook::velox::functions::qdigest::QuantileDigest<T, StlAllocator<T>>
+      digest_;
+};
+
+template <typename T>
+class MergeQDigestAggregate : public exec::Aggregate {
+ public:
+  explicit MergeQDigestAggregate(const TypePtr& resultType)
+      : exec::Aggregate(resultType) {}
+
+  int32_t accumulatorFixedWidthSize() const override {
+    return sizeof(QDigestAccumulator<T>);
+  }
+
+  int32_t accumulatorAlignmentSize() const override {
+    return alignof(QDigestAccumulator<T>);
+  }
+
+  bool accumulatorUsesExternalMemory() const override {
+    return true;
+  }
+
+  bool isFixedSize() const override {
+    return false;
+  }
+
+  void toIntermediate(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      VectorPtr& result) const final {
+    singleInputAsIntermediate(rows, args, result);
+  }
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    extractAccumulators(groups, numGroups, result);
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    VELOX_CHECK(result);
+    auto* flatResult = (*result)->asFlatVector<StringView>();
+    extract(groups, numGroups, flatResult);
+  }
+
+  void addRawInput(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushdown*/) override {
+    addIntermediateResults(groups, rows, args, false /*unused*/);
+  }
+
+  void addIntermediateResults(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushdown*/) override {
+    decodedQDigest_.decode(*args[0], rows, true);
+
+    rows.applyToSelected([&](auto row) {
+      if (decodedQDigest_.isNullAt(row)) {
+        return;
+      }
+
+      auto group = groups[row];
+      auto tracker = trackRowSize(group);
+      clearNull(group);
+
+      mergeToAccumulator(group, row);
+    });
+  }
+
+  void addSingleGroupRawInput(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushdown*/) override {
+    auto tracker = trackRowSize(group);
```

Comment:
This is already called in in addSingleGroupIntermediateResults(), so not needed here.